### ISSUE TITLE
Exporting trained model

### ISSUE THREAD

**Issue by periscopiojoe** [menpo/menpo#808]
Hello,
I just can´t finish my project because I  I can´t   export my trained model.I am working with Miniconda and python 2.7, I tried this: 

> import menpo.io as mio
> menpo.io.export_pickle(aam, '/Users/usuario/Desktop/test/aammtrained.pkl', overwrite=False, protocol=2)

**Error:  name "menpo" is not defined** 

In other side, when I load the training images, just 35 are allowed, I used this function:

 > training_images = imagestrain.map(load_image)

**Error: "All max indices must be greater that the min indices"**
Is there another possibility to load more images?

I have been working hard for months with your cool api , but i need this help to finish it, please ;-( 
I already readed the docs, but I couldn´t figure out it.
Thanks

**Comment by jabooth** [comment]
mio.export_pickle not menpo.io.export_pickle

As for the second, I'd need more context. There is no arbitrary limit on
the number of images you can import in menpo.

**Comment by periscopiojoe** [comment]
[FINAL CODE.zip](https://github.com/menpo/menpo/files/1118761/FINAL.CODE.zip)
Thank you James,
he re you are attached my code, i can work wit just  35 images in thew traines, whe I tried to load 100 more  I always get the same error, I reviewed as well every single image and landmark, and everything is ok,,,

**Comment by jabooth** [comment]
Can you post the full stack trace with the second error?

(the 'All max indices must be greater that the min indices' one)

**Comment by periscopiojoe** [comment]
C:\Users\usuario\Miniconda3\envs\envjose\lib\site-packages\menpo\image\base.py:1157:
RuntimeWarning: invalid value encountered in greater
  elif not np.all(max_indices > min_indices):

---------------------------------------------------------------------------ValueError
                               Traceback (most recent call
last)<ipython-input-28-ffa943b8a32a> in <module>()      6
diagonal=200,      7     patch_shape=[(19, 19), (17, 17), (15,
15)],----> 8     scales=(0.25, 0.5, 1.0)      9 )     10
~\Miniconda3\envs\envjose\lib\site-packages\menpofit\aam\base.py in
__init__(self, images, group, holistic_features, reference_shape,
diagonal, scales, patch_shape, patch_normalisation, shape_model_cls,
max_shape_components, max_appearance_components, verbose, batch_size)
 1137             max_shape_components=max_shape_components,   1138
         max_appearance_components=max_appearance_components,-> 1139
          shape_model_cls=shape_model_cls, batch_size=batch_size)
1140    1141     @property
~\Miniconda3\envs\envjose\lib\site-packages\menpofit\aam\base.py in
__init__(self, images, group, holistic_features, reference_shape,
diagonal, scales, transform, shape_model_cls, max_shape_components,
max_appearance_components, verbose, batch_size)    137         # Train
AAM    138         self._train(images, increment=False, group=group,
verbose=verbose,--> 139                     batch_size=batch_size)
140     141     def _train(self, images, increment=False, group=None,
~\Miniconda3\envs\envjose\lib\site-packages\menpofit\aam\base.py in
_train(self, images, increment, group, shape_forgetting_factor,
appearance_forgetting_factor, verbose, batch_size)    149
image_batches = batch(images, batch_size)    150         else:--> 151
           image_batches = [list(images)]    152     153         for
k, image_batch in enumerate(image_batches):
~\Miniconda3\envs\envjose\lib\_collections_abc.py in __iter__(self)
815         try:    816             while True:--> 817
v = self[i]    818                 yield v    819                 i +=
1
~\Miniconda3\envs\envjose\lib\site-packages\menpo\base.py in
__getitem__(self, slice_)    495         elif isinstance(slice_, int)
or hasattr(slice_, '__index__'):    496             # PEP 357 and
single integer index access - returns element--> 497
return self._callables[slice_]()    498         else:    499
  # A slice or unknown type is passed - let List handle it
~\Miniconda3\envs\envjose\lib\site-packages\menpo\base.py in
delayed(delay_f, delay_x)    583         # is passed the actual
instantiated object and not the callable itself.    584         def
delayed(delay_f, delay_x):--> 585             return
delay_f(delay_x())    586     587         if isinstance(f,
collections.Iterable) and callable(f):
<ipython-input-26-8478542a9041> in load_image(i)      2 # and turn to
greyscale - again to save memory      3 def load_image(i):----> 4
i, tr = i.crop_to_landmarks_proportion(0.3, return_transform=True)
 5     if i.n_channels == 3:      6         i = i.as_greyscale()
~\Miniconda3\envs\envjose\lib\site-packages\menpo\image\base.py in
crop_to_landmarks_proportion(self, boundary_proportion, group,
minimum, constrain_to_boundary, return_transform)   1355
pc, boundary_proportion, minimum=minimum,   1356
constrain_to_boundary=constrain_to_boundary,-> 1357
return_transform=return_transform)   1358    1359     def
constrain_points_to_bounds(self, points):
~\Miniconda3\envs\envjose\lib\site-packages\menpo\image\base.py in
crop_to_pointcloud_proportion(self, pointcloud, boundary_proportion,
minimum, constrain_to_boundary, return_transform)   1304
pointcloud, boundary=boundary,   1305
constrain_to_boundary=constrain_to_boundary,-> 1306
return_transform=return_transform)   1307    1308     def
crop_to_landmarks_proportion(self, boundary_proportion,
~\Miniconda3\envs\envjose\lib\site-packages\menpo\image\base.py in
crop_to_pointcloud(self, pointcloud, boundary, constrain_to_boundary,
return_transform)   1210         return self.crop(min_indices,
max_indices,   1211
constrain_to_boundary=constrain_to_boundary,-> 1212
      return_transform=return_transform)   1213    1214     def
crop_to_landmarks(self, group=None, boundary=0,
~\Miniconda3\envs\envjose\lib\site-packages\menpo\image\base.py in
crop(self, min_indices, max_indices, constrain_to_boundary,
return_transform)   1156                 " length n_dims
({})".format(self.n_dims))   1157         elif not np.all(max_indices
> min_indices):-> 1158             raise ValueError("All max indices
must be greater that the min "   1159
"indices")   1160         min_bounded =
self.constrain_points_to_bounds(min_indices)
ValueError: All max indices must be greater that the min indices



2017-07-11 17:54 GMT+02:00 James Booth <notifications@github.com>:

**Comment by jabooth** [comment]
With markdown:
```
C:\Users\usuario\Miniconda3\envs\envjose\lib\site-packages\menpo\image\base.py:1157:
RuntimeWarning: invalid value encountered in greater
  elif not np.all(max_indices > min_indices):

---------------------------------------------------------------------------ValueError
                               Traceback (most recent call
last)<ipython-input-28-ffa943b8a32a> in <module>()      6
diagonal=200,      7     patch_shape=[(19, 19), (17, 17), (15,
15)],----> 8     scales=(0.25, 0.5, 1.0)      9 )     10
~\Miniconda3\envs\envjose\lib\site-packages\menpofit\aam\base.py in
__init__(self, images, group, holistic_features, reference_shape,
diagonal, scales, patch_shape, patch_normalisation, shape_model_cls,
max_shape_components, max_appearance_components, verbose, batch_size)
 1137             max_shape_components=max_shape_components,   1138
         max_appearance_components=max_appearance_components,-> 1139
          shape_model_cls=shape_model_cls, batch_size=batch_size)
1140    1141     @property
~\Miniconda3\envs\envjose\lib\site-packages\menpofit\aam\base.py in
__init__(self, images, group, holistic_features, reference_shape,
diagonal, scales, transform, shape_model_cls, max_shape_components,
max_appearance_components, verbose, batch_size)    137         # Train
AAM    138         self._train(images, increment=False, group=group,
verbose=verbose,--> 139                     batch_size=batch_size)
140     141     def _train(self, images, increment=False, group=None,
~\Miniconda3\envs\envjose\lib\site-packages\menpofit\aam\base.py in
_train(self, images, increment, group, shape_forgetting_factor,
appearance_forgetting_factor, verbose, batch_size)    149
image_batches = batch(images, batch_size)    150         else:--> 151
           image_batches = [list(images)]    152     153         for
k, image_batch in enumerate(image_batches):
~\Miniconda3\envs\envjose\lib\_collections_abc.py in __iter__(self)
815         try:    816             while True:--> 817
v = self[i]    818                 yield v    819                 i +=
1
~\Miniconda3\envs\envjose\lib\site-packages\menpo\base.py in
__getitem__(self, slice_)    495         elif isinstance(slice_, int)
or hasattr(slice_, '__index__'):    496             # PEP 357 and
single integer index access - returns element--> 497
return self._callables[slice_]()    498         else:    499
  # A slice or unknown type is passed - let List handle it
~\Miniconda3\envs\envjose\lib\site-packages\menpo\base.py in
delayed(delay_f, delay_x)    583         # is passed the actual
instantiated object and not the callable itself.    584         def
delayed(delay_f, delay_x):--> 585             return
delay_f(delay_x())    586     587         if isinstance(f,
collections.Iterable) and callable(f):
<ipython-input-26-8478542a9041> in load_image(i)      2 # and turn to
greyscale - again to save memory      3 def load_image(i):----> 4
i, tr = i.crop_to_landmarks_proportion(0.3, return_transform=True)
 5     if i.n_channels == 3:      6         i = i.as_greyscale()
~\Miniconda3\envs\envjose\lib\site-packages\menpo\image\base.py in
crop_to_landmarks_proportion(self, boundary_proportion, group,
minimum, constrain_to_boundary, return_transform)   1355
pc, boundary_proportion, minimum=minimum,   1356
constrain_to_boundary=constrain_to_boundary,-> 1357
return_transform=return_transform)   1358    1359     def
constrain_points_to_bounds(self, points):
~\Miniconda3\envs\envjose\lib\site-packages\menpo\image\base.py in
crop_to_pointcloud_proportion(self, pointcloud, boundary_proportion,
minimum, constrain_to_boundary, return_transform)   1304
pointcloud, boundary=boundary,   1305
constrain_to_boundary=constrain_to_boundary,-> 1306
return_transform=return_transform)   1307    1308     def
crop_to_landmarks_proportion(self, boundary_proportion,
~\Miniconda3\envs\envjose\lib\site-packages\menpo\image\base.py in
crop_to_pointcloud(self, pointcloud, boundary, constrain_to_boundary,
return_transform)   1210         return self.crop(min_indices,
max_indices,   1211
constrain_to_boundary=constrain_to_boundary,-> 1212
      return_transform=return_transform)   1213    1214     def
crop_to_landmarks(self, group=None, boundary=0,
~\Miniconda3\envs\envjose\lib\site-packages\menpo\image\base.py in
crop(self, min_indices, max_indices, constrain_to_boundary,
return_transform)   1156                 " length n_dims
({})".format(self.n_dims))   1157         elif not np.all(max_indices
> min_indices):-> 1158             raise ValueError("All max indices
must be greater that the min "   1159
"indices")   1160         min_bounded =
self.constrain_points_to_bounds(min_indices)
ValueError: All max indices must be greater that the min indices
```

**Comment by jabooth** [comment]
@periscopiojoe can you please put the stack trace directly on Github and not by responding to email - that trashes the indentation making it very hard to follow.

**Comment by patricksnape** [comment]
Looks like two points are coincident which suggest a zero size bounding box.

**Comment by periscopiojoe** [comment]
I am sorry @jabooth, here you are:

`C:\Users\usuario\Miniconda3\envs\envjose\lib\site-packages\menpo\image\base.py:1157: RuntimeWarning: invalid value encountered in greater
  elif not np.all(max_indices > min_indices):

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-28-ffa943b8a32a> in <module>()
      6     diagonal=200,
      7     patch_shape=[(19, 19), (17, 17), (15, 15)],
----> 8     scales=(0.25, 0.5, 1.0)
      9 )
     10 

~\Miniconda3\envs\envjose\lib\site-packages\menpofit\aam\base.py in __init__(self, images, group, holistic_features, reference_shape, diagonal, scales, patch_shape, patch_normalisation, shape_model_cls, max_shape_components, max_appearance_components, verbose, batch_size)
   1137             max_shape_components=max_shape_components,
   1138             max_appearance_components=max_appearance_components,
-> 1139             shape_model_cls=shape_model_cls, batch_size=batch_size)
   1140 
   1141     @property

~\Miniconda3\envs\envjose\lib\site-packages\menpofit\aam\base.py in __init__(self, images, group, holistic_features, reference_shape, diagonal, scales, transform, shape_model_cls, max_shape_components, max_appearance_components, verbose, batch_size)
    137         # Train AAM
    138         self._train(images, increment=False, group=group, verbose=verbose,
--> 139                     batch_size=batch_size)
    140 
    141     def _train(self, images, increment=False, group=None,

~\Miniconda3\envs\envjose\lib\site-packages\menpofit\aam\base.py in _train(self, images, increment, group, shape_forgetting_factor, appearance_forgetting_factor, verbose, batch_size)
    149             image_batches = batch(images, batch_size)
    150         else:
--> 151             image_batches = [list(images)]
    152 
    153         for k, image_batch in enumerate(image_batches):

~\Miniconda3\envs\envjose\lib\_collections_abc.py in __iter__(self)
    815         try:
    816             while True:
--> 817                 v = self[i]
    818                 yield v
    819                 i += 1

~\Miniconda3\envs\envjose\lib\site-packages\menpo\base.py in __getitem__(self, slice_)
    495         elif isinstance(slice_, int) or hasattr(slice_, '__index__'):
    496             # PEP 357 and single integer index access - returns element
--> 497             return self._callables[slice_]()
    498         else:
    499             # A slice or unknown type is passed - let List handle it

~\Miniconda3\envs\envjose\lib\site-packages\menpo\base.py in delayed(delay_f, delay_x)
    583         # is passed the actual instantiated object and not the callable itself.
    584         def delayed(delay_f, delay_x):
--> 585             return delay_f(delay_x())
    586 
    587         if isinstance(f, collections.Iterable) and callable(f):

<ipython-input-26-8478542a9041> in load_image(i)
      2 # and turn to greyscale - again to save memory
      3 def load_image(i):
----> 4     i, tr = i.crop_to_landmarks_proportion(0.3, return_transform=True)
      5     if i.n_channels == 3:
      6         i = i.as_greyscale()

~\Miniconda3\envs\envjose\lib\site-packages\menpo\image\base.py in crop_to_landmarks_proportion(self, boundary_proportion, group, minimum, constrain_to_boundary, return_transform)
   1355             pc, boundary_proportion, minimum=minimum,
   1356             constrain_to_boundary=constrain_to_boundary,
-> 1357             return_transform=return_transform)
   1358 
   1359     def constrain_points_to_bounds(self, points):

~\Miniconda3\envs\envjose\lib\site-packages\menpo\image\base.py in crop_to_pointcloud_proportion(self, pointcloud, boundary_proportion, minimum, constrain_to_boundary, return_transform)
   1304             pointcloud, boundary=boundary,
   1305             constrain_to_boundary=constrain_to_boundary,
-> 1306             return_transform=return_transform)
   1307 
   1308     def crop_to_landmarks_proportion(self, boundary_proportion,

~\Miniconda3\envs\envjose\lib\site-packages\menpo\image\base.py in crop_to_pointcloud(self, pointcloud, boundary, constrain_to_boundary, return_transform)
   1210         return self.crop(min_indices, max_indices,
   1211                          constrain_to_boundary=constrain_to_boundary,
-> 1212                          return_transform=return_transform)
   1213 
   1214     def crop_to_landmarks(self, group=None, boundary=0,

~\Miniconda3\envs\envjose\lib\site-packages\menpo\image\base.py in crop(self, min_indices, max_indices, constrain_to_boundary, return_transform)
   1156                 " length n_dims ({})".format(self.n_dims))
   1157         elif not np.all(max_indices > min_indices):
-> 1158             raise ValueError("All max indices must be greater that the min "
   1159                              "indices")
   1160         min_bounded = self.constrain_points_to_bounds(min_indices)

ValueError: All max indices must be greater that the min indices

`

**Comment by periscopiojoe** [comment]
Yesterday It allowed me to load 59 images for training, no more, I can´t understand this behaviour...
thanks for your replys guys.

**Comment by periscopiojoe** [comment]
so what do you think @jabooth @patricksnape? Does the mistake come from "map load" or "bounding box" ?
I can' t continue, please help :(
Thank you in advance!

**Comment by patricksnape** [comment]
I think the mistake comes from the fact that one of your bounding boxes is incorrect in the input data.

**Comment by periscopiojoe** [comment]
thank you very much guys!
